README.md

```diff
@@ -1,17 +1,139 @@
-# UCAN IPLD Schema
+# UCAN IPLD Schema v0.1.0
 
 ## Editors
 
-* _
+* [Irakli Gozalishvili](https://github.com/Gozala), DAG House
 
 ## Authors
 
-* _
+* [Hugo Dias](https://github.com/hugomrdias), DAG House
+* [Irakli Gozalishvili](https://github.com/Gozala), DAG House
+* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)
+
+## Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
 # 0 Abstract
 
-## Language
+[Interplanetary Linked Data (IPLD)](https://ipld.io/) is a consistent, highly general data model for content addressed linked data forming any DAG. This data model provides a convenient pivot between many serializations of the same information. Being a chained token model, UCANs can be very naturally expressed via IPLD. However, JWTs are not fully deterministic due to whitespace and key ordering. This specification defines an IPLD Schema for UCAN, and a canonical JWT serialization for compatibility with all other clients.
+
+# 1 Motivation
+
+The [core UCAN specification](https://github.com/ucan-wg/spec) is defined as a JWT to make it easily adoptable with existing tools, and as a common format for all implementations. There are many cases where different encodings are preferred for reasons such as compactness, machine-efficient formats, and so on. This specification outlines a format based on IPLD that can be deterministically encoded and decoded between many serialization formats, while still being able to encode as JWT for compatibility.
+
+# 2 IPLD Schema
+
+Unlike a JWT, the IPLD encoding of UCAN does not require separate header, claims, and signature fields.
+
+```ipldsch
+type UCAN struct {
+  ucv String 
+
+  iss DID
+  aud DID 
+  sig Signature
+
+  att [Capability] 
+  prf [&UCAN]
+  exp Int
+
+  fct [Fact]
+  nnc optional String
+  nbf optional Int
+} representation map {
+  field fct default []
+  field prf default []
+}
+```
+
+## 2.1 DID
+
+DIDs MUST be encoded as [DID](https://www.w3.org/TR/did-core/)s. The [`did:key` method](https://w3c-ccg.github.io/did-method-key/) is RECOMMENDED as it is self-contained.
+
+``` ipldsch
+type DID struct {
+  m String # Method
+  i String # Method-specific identifier
+}
```

Comment:
This isn't what I meant, I was merely suggesting `type DID = Bytes` with a following multicodec codes

| code     |  kind      |
|-------- |---------|
| 0xed     | Ed25519 `did:key:` |
| 0x1205 | RSA `did:key:`         |
| ???        | `did:ion:`                 |


I don't know maybe that is an overkill and `type DID = String` is good enough. I just felt that it was silly to waste bytes on `did:key:` prefix and we could extend above table with more codes as needed in order to avoid wasting bytes.

If UCAN library can not turn `iss` and `aud` into something like:

```ts
interface Principal {
   did(): `did:${string}`
   verify(payload:Uint8Array): Promise<boolean>
}
```

There is no practical difference between it been a string or some bytes, apart from printing it out I suppose.

Reply:
Ok it seems like we should be able to use [byteprefixed unions](https://ipld.io/docs/schemas/features/representation-strategies/#union-bytesprefix-representation)

```ipldsch
type Principal union {
  | Ed25519 "0xed"
  | RSA "0x1205"
  | DID "0x64"
} representation bytesprefix
```

Reply:
> <p dir="auto">I don't know maybe that is an overkill and <code class="notranslate">type DID = String</code> is good enough. I just felt that it was silly to waste bytes on <code class="notranslate">did:key:</code> prefix and we could extend above table with more codes as needed in order to avoid wasting bytes.</p>

Oh I see @Gozala. These are actually slightly different axes. A arbitrary DID could contain a Ed25519 key. A really good example of this is `did:pkh`, which is like `did:key` but has both a key _and_ some information about which blockchain it's intended for (since many of them share the same elliptic curve).

🤔 You are right that `did:key` is "just" bytes. Adding the tag ahead like in my proposal has some extra bytes, but may give us some flexibility. What about something like this:

```ipldscheme
type DID struct {
  i Bytes # Method-specific identifier
  m optional String # Method
  # ^^^^^^^^
}
```

Where we can omit the tag on `did:key`s, because they're roughly an identity DID. Less to store at least. Thoughts?

Reply:
Ah, which is "just" a different layout from byteprefixed. I guess that union would be arbitrarily long, since we could handle anything in `did:key` 🤔

Would you want a separate `did:ion`, `did:pkh`, `did:3`, etc prefix list somewhere, or would if be something like:

```
type DID struct {
  m String
  i String
}

type Principal union {
  | Ed25519 "0xed"
  | RSA "0x1205"
  | DID "0x64"
} representation bytesprefix
```

```js
did_ion = append(0x64, DID(ION, ionId).serialize())
```

...roughly?

Reply:
> Would you want a separate did:ion, did:pkh, did:3, etc prefix list somewhere, or would if be something like:

No what I meant is that UTF8 serialization of arbitrary did will start with `0x64` byte so last variant is catch all for all the possible DIDs. We could further expand `Principal` union as needed to remove need for encoding did prefixes when they are common enough to matter. e.g. if `did:ion:` becomes widely used we could add `| ION 0x65`.

Reply:
It could be that added complexity here does not warrant shaving a few bytes, but then again this is what spec says

> Implementations MUST include the [did:key method](https://w3c-ccg.github.io/did-method-key/), and MAY be augmented with [additional DID methods](https://www.w3.org/TR/did-core/).

Which somewhat guides me to optimize primary case and make extensions possible. Which is kind of why originally principals were defined as:

```ipldsch
-- multicodec tagged public key
-- 0xed         Ed25519
-- 0x1205     RSA
type Principal = Bytes
```

This optimizes for spec, while still allowing for arbitrary extensions. You could literary encode `did:ion:...` into UTF8 and call it a day or you could optimize specific cases you care about as well. From the implementation perspective, I can't really do anything useful with non `did:key` DIDs unless you plug your own "principal encoder/decoder" and when you do I don't care what bytes it would output or intake.

If this sounds controversial, I'm ok with just strings. I'm just not sure what the benefits of using strings are. Only compelling argument I've heard so far is that this makes `did:key`s 1st class citizen and all the other DIDs 2nd class, but then again I feel like this is aligned with a spec and it seems reasonable to optimize for common case.

If spec ends up supporting more did's natively we could always expand multicodec table to optimize those cases.

Reply:
One additional problem with non did:key's is that when encoding them into JWT it is no longer clear what signature algorithm should be. E.g. consider following case:

issuer of the UCAN is `did:dns:danubetech.com` which is e.g. resolves as:

```
_key1._did.danubetech.com. IN URI 100 10 "did:key:z6MkjvBkt8ETnxXGBFPSGgYKb43q7oNHLX8BiYSPcXVG6gY6"
```

From the current data model one could not deduce without out of bound IO that jwt algorithm is EdDSA. In other words we can not take such IPLD ucan and serialize into JWT as we won't be able to derive `alg` of the header.

Even if we add additional field to capture that info we face another problem. What if `did:dns:danubetech.com` record has changed to point to say RSA key in the meantime ? UCAN now becomes invalid (although argument can be made that this is desired).

P.S: We have considered alternative approach for our use cases. That is, allowing use of non `did:key:`s in `iss` / `aud` but only if corresponding DID doc is provided in `fct`s .

Reply:
> as we won't be able to derive alg of the header

Ooh you're hoping to skip the `alg` as well 🤔 I get that this is redundant in the `did:key` case, but not in the general case. Even with e.g. a `did:key` secp256k1, you can perform both either a ECDSA or Ethereum signature (which we're running into now that we're adding MetaMask support).

> Only compelling argument I've heard so far is that this makes did:keys 1st class citizen and all the other DIDs 2nd class, but then again I feel like this is aligned with a spec and it seems reasonable to optimize for common case.

Yeah, this is roughly my thought process as well. While I also have the gut reaction about making other DIDs second class, optimizing for the common case makes a ton of sense to me.

I'll do another revision and see what you think 🚀

Reply:
I saw the comment about `multsignature` in https://github.com/ucan-wg/ucan-ipld/issues/4. That's an interesting idea, and may be useful elsewhere too. Does such a thing exist today, or would it be something we extract out of this spec later?

Reply:
<img width="767" alt="Screen Shot 2022-09-12 at 17 16 47" src="https://user-images.githubusercontent.com/1052016/189780481-e4bd5a18-0b5b-4316-82be-1ed245a9e3ce.png">

Looks like there's a couple of these in multicodec already... but only for Holochain? 🤔